### PR TITLE
DM-41366: Add Docker image build for butler server

### PR DESCRIPTION
**Pull request by dhirving**
## Checklist

- [x] ran Jenkins
- [ ] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+RUN pip install --quiet --no-cache-dir -r requirements.txt -r server_requirements.txt
+
+FROM dependencies-image AS install-image
+
+# Use the virtualenv
+ENV PATH="/opt/venv/bin:$PATH"
+
+COPY . /workdir
+WORKDIR /workdir
+RUN pip install --no-cache-dir .
```
Comment (timj) [inline Dockerfile]: Maybe add `--no-deps` here so it doesn't try again to install all the dependencies that we've installed via `requirements.txt` above? Historically we have had some weird things going on with version number matches if requirements.txt is pointing to git repos and pip is using `pyproject.toml` dependencies from PyPI.

File: compose.yml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+# test data from the ci_hsc_gen3 package.  Assumes that ci_hsc_gen3
+# has been built via lsstsw in a directory adjacent to the directory
+# containing this file.
+services:
+  butler:
+    build: .
+    ports:
+      - "8080:8080"
+    environment:
+      BUTLER_SERVER_BUTLER_ROOT: "/butler_root"
```
Comment (timj) [inline compose.yml]: I am pondering whether BUTLER_ROOT is the right term now. This is really the URI to the butler configuration (with or without the `butler.yaml`).

File: server/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,2 @@
+uvicorn
+psycopg2
```
Comment (timj) [inline server/requirements.txt:2]: In theory `pip install .[postgres]` should pick up this dependency (it's listed as an optional dep in the pyproject.toml), but not if I've just told you to use `--no-deps` for that part above...

**Review by timj — APPROVED** [review]
Looks good. It would be nice if the github action could run a simple test to check that the code can be loaded. eg by running something like:

```bash
$ butler create tmp
$ butler config-dump tmp
```

File: server/requirements.txt (hunk 3 above)
Comment (dhirving) [reply server/requirements.txt:2]: The dependencies actually want to be installed in a separate step from the pip install of the application itself -- if the requirements.txt files don't change, Docker can cache the intermediate dependency image and save a few minutes on the next build.

Also the requirements will need to be pinned in future so we won't want pyproject.toml involved.

File: compose.yml (hunk 2 above)
Comment (dhirving) [reply compose.yml]: renamed BUTLER_SERVER_CONFIG_URI

**Pull request merged**